### ISSUE TITLE
Webnovel Guard

### ISSUE THREAD

**Issue by C0rn3j** [mHaisham/webnovelbot#1]
```python
webnovel = WebnovelBot(timeout=360)
NOVEL = 'https://www.webnovel.com/book/book-name_123456789'
webnovel.driver.get(NOVEL)

novel_id = webnovel.novel_id

webnovel.signin(email, password)

claim_daily(webnovel, NOVEL)
```

Sometimes guard is required, which it looks like the library doesn't handle(I also had to do reCaptcha in this example) and will redirect to:

```
https://passport.webnovel.com/guard.htmlhttps://passport.webnovel.com/emaillogin.html&sessionKey=abcdef&encry=longkeywithlotsof%s&email=myemail
```

EDIT: And actually if I do get through the guard, I get stuck anyways

```python
    claim_daily(webnovel, NOVEL)
  File "/home/c0rn3j/test.py", line 12, in claim_daily
    bot.claim_tasks()
  File "/home/c0rn3j/.local/lib/python3.9/site-packages/webnovel/decorators/signin.py", line 16, in __call__
    return self.__wrapped__(self.__self__, *args, **kwargs)
  File "/home/c0rn3j/.local/lib/python3.9/site-packages/webnovel/bot.py", line 453, in claim_tasks
    self._focus_profile()
  File "/home/c0rn3j/.local/lib/python3.9/site-packages/webnovel/bot.py", line 507, in _focus_profile
    profile_button = self.driver.find_element_by_css_selector("div[class^='g_user'][class*='_hover']")
  File "/home/c0rn3j/.local/lib/python3.9/site-packages/selenium/webdriver/remote/webdriver.py", line 598, in find_element_by_css_selector
    return self.find_element(by=By.CSS_SELECTOR, value=css_selector)
  File "/home/c0rn3j/.local/lib/python3.9/site-packages/selenium/webdriver/remote/webdriver.py", line 976, in find_element
    return self.execute(Command.FIND_ELEMENT, {
  File "/home/c0rn3j/.local/lib/python3.9/site-packages/selenium/webdriver/remote/webdriver.py", line 321, in execute
    self.error_handler.check_response(response)
  File "/home/c0rn3j/.local/lib/python3.9/site-packages/selenium/webdriver/remote/errorhandler.py", line 242, in check_response
    raise exception_class(message, screen, stacktrace)
selenium.common.exceptions.NoSuchElementException: Message: no such element: Unable to locate element: {"method":"css selector","selector":"div[class^='g_user'][class*='_hover']"}
  (Session info: chrome=88.0.4324.182)
```

claim_daily is as per example, just reordered.

```python

def claim_daily(bot, novel=None):
	# claim daily fast pass
	bot.claim_tasks()
	bot.energy_vote([1])
	bot.power_vote()

	import time
	time.sleep(5)

	if novel is not None:
		bot.driver.get(novel)
```

- [x] Email guard support
- [x] Fix daily FP claim
- [x] Fix power stone use
- [x] Fix profile selection
- [x] Click away preferences window

**Comment by C0rn3j** [comment]
I think this is because the button changed (or A/B testing is happening?) and the profile button now has `g_dropdown` instead of `_hover`:
`g_user j_header_user g_dropdown _on` (_on is there since I was currently hovering it)

---

Next up, I always get this pop up on new login (I guess it saves the answer in cookies since it keeps pestering me) and it does not look like that one is being handled.
![image](https://user-images.githubusercontent.com/1641362/108863456-868eb680-75f1-11eb-8c38-f30e6ea4f387.png)

Looks like the confirm button is this:
`<button type="submit" title="confirm" class="bt _m mw160 j_post_preference">confirm</button>`

So I did the following, which seems to work -
  * added a new function to click away the preferences (this could be done better I suppose):
```python
    def yeetPreferences(self):
        """
        click the preferences button
        """
        signin_btn = self.driver.find_element_by_class_name('j_post_preference') # bt _m mw160 j_post_preference
        signin_btn.click()
```

  * modified the profile function to use the new CSS and also to click instead of just hover, as that doesn't do anything anymore
```python
    def _focus_profile(self):
        """
        moves mouse over to profile button

        :return: profile button element
        """
        profile_button = self.driver.find_element_by_css_selector("div[class^='g_user'][class*='g_dropdown']")
        ActionChains(self.driver).click(profile_button).perform()

        return profile_button
```

Now there's two remaining issues - 

1. Daily claim does not work
2. Power voting does not work. I believe this is because I get redirected to the voting page or simply another page other than the defined web novel, the function does not try to load the novel page if it differs

**Comment by m-haisham** [comment]
I've been working on my other projects an as a result have neglected this.

would you be doing a pull request?

**Comment by C0rn3j** [comment]
(I updated my above comment while you responded)

```
        for claim in self.driver.find_elements_by_css_selector('.j_claim_task'):
            claim.click()
```
Is this code not triggering on the following page because the class contains more elements?
```
<a href="javascript:" class="bt task-btn _warning j_claim_task" data-task-id="100101" data-report-eid="qi_K05">claim</a>
```

This is my first time using Selenium and I'm not a web developer on top so I am a bit out of my depth here, I've no issue sending a PR when I get it all working

**Comment by m-haisham** [comment]
I plan to add the usage of existing browser cookies [`chrome`, `firefox`] using `browser_cookies3` package, you can expect it within a few hours

This should hopefully help forgo the troubles that the login process brings

after that i will add the suggestions you have given, unless you create a pull request

**Comment by mHaisham** [comment]
I plan to add the usage of existing browser cookies [`chrome`, `firefox`] using `browser_cookies3` package, you can expect it within a few hours

This should hopefully help forgo the troubles that the login process brings

after that i will add the suggestions you have given, unless you create a pull request

**Comment by C0rn3j** [comment]
I sent a PR for preferences and broken profile function.

I can't seem to figure out how the daily FP claim button broke, and I didn't look into the power stone issue yet (though I am 90% certain it's due to the current URL not being checked), so if you'll have time to look into it yourself, that'd be great!

**Comment by m-haisham** [comment]
I've looked through the daily FP and power stones buttons, the selectors seem to checkout. I havent done any testing yet.

**Comment by m-haisham** [comment]
Unfortunately this is all the time i have for today. I'll test them tomorrow

**Comment by C0rn3j** [comment]
The issue is with claim_tasks (daily FP) is the waiting part, it seems it does not wait til it's fully loaded.

Or at least I'm assuming that since I added 1 second sleeps between each part and it worked this time.

---

About power_vote  - one issue is in the provided example, it should have `bot.power_vote(novel)` instead of `bot.power_vote()`

Another is that it seems it does not wait til the page fully loads either. Worked around it with a sleep.

And thirdly, it seems to error out if the power_vote daily is already claimed

```python
Traceback (most recent call last):
  File "/home/c0rn3j/test.py", line 34, in <module>
    claim_daily(webnovel, NOVEL)
  File "/home/c0rn3j/test.py", line 14, in claim_daily
    bot.power_vote(novel)
  File "/home/c0rn3j/.local/lib/python3.9/site-packages/webnovel/decorators/signin.py", line 16, in __call__
    return self.__wrapped__(self.__self__, *args, **kwargs)
  File "/home/c0rn3j/.local/lib/python3.9/site-packages/webnovel/bot.py", line 538, in power_vote
    chain.perform()
  File "/home/c0rn3j/.local/lib/python3.9/site-packages/selenium/webdriver/common/action_chains.py", line 80, in perform
    self.w3c_actions.perform()
  File "/home/c0rn3j/.local/lib/python3.9/site-packages/selenium/webdriver/common/actions/action_builder.py", line 76, in perform
    self.driver.execute(Command.W3C_ACTIONS, enc)
  File "/home/c0rn3j/.local/lib/python3.9/site-packages/selenium/webdriver/remote/webdriver.py", line 321, in execute
    self.error_handler.check_response(response)
  File "/home/c0rn3j/.local/lib/python3.9/site-packages/selenium/webdriver/remote/errorhandler.py", line 242, in check_response
    raise exception_class(message, screen, stacktrace)
selenium.common.exceptions.JavascriptException: Message: javascript error: Failed to execute 'elementsFromPoint' on 'Document': The provided double value is non-finite.
  (Session info: chrome=88.0.4324.182)
```

**Comment by m-haisham** [comment]
Alright. I'll fix these asap.

**Comment by m-haisham** [comment]
commit 06adb11002161177463f1e2dc9004a1422e368fd should fix both power stone and claim daily issue

I am unfortunately unable to trigger the guard

**Comment by C0rn3j** [comment]
>I am unfortunately unable to trigger the guard

I had it trigger on about every second new account when I was testing.

If you give me somewhere to privately contact you, I can PM you a list you can run through which should trigger it.

**Comment by m-haisham** [comment]
you can send to the temp mail `bievog@fastness.work`

**Comment by m-haisham** [comment]
alright. i got it

**Comment by m-haisham** [comment]
pushed commit 91d77962eb8f20095a4a483076f41bcba3179b6d to fix the guard issue

as to my understanding guard appears when the user is not found in the webnovel database.
hence I've made it throw a `ValueError` when guard appears

```python
try:
    webnovel.signin(USER_EMAIL, USER_PASS)
except ValueError:
   # do something on sign in failure
```

**Comment by C0rn3j** [comment]
>as to my understanding guard appears when the user is not found in the webnovel database.

That's not exactly it, it's email 2FA that triggers under some unknown conditions, usually new 'device'.
![image](https://user-images.githubusercontent.com/1641362/108995492-79c79c80-769d-11eb-8987-931bc82229ee.png)


```python
Traceback (most recent call last):
  File "/home/c0rn3j/test.py", line 39, in <module>
    webnovel.signin(account, password)
  File "/home/c0rn3j/.local/lib/python3.9/site-packages/webnovel/bot.py", line 128, in signin
    raise ValueError('user not found')
ValueError: user not found
```
That is indeed catching it, but wouldn't it be better to redirect to the form where the code can be entered, so user can interact and fill it in?

Currently this happens if reCAPTCHA gets triggered, the bot simply waits til timeout until captcha is filled out and login is button is pressed again.

---

Side note, `pip install git+https://github.com/mHaisham/webnovelbot.git` fails if the dependencies in requirements.txt aren't already installed (the cookie dependency in my case).

**Comment by m-haisham** [comment]
> but wouldn't it be better to redirect to the form where the code can be entered, so user can interact and fill it in?

I'll add make this default and single attempt optional

> Currently this happens if reCAPTCHA gets triggered

certainly something that can be handled better

> fails if the dependencies in requirements.txt aren't already installed

Its a problem with browser-cookie3. i'll add a warning to README

**Comment by m-haisham** [comment]
pushed commit 9f74df755c7ab65f5013b289e5dc5798a0639572

this gives better support handling captcha and guard.

Read more on [README.md](https://github.com/mHaisham/webnovelbot#signin)

**Comment by C0rn3j** [comment]
That's perfect!
There seems to be a small issue with it though
```python
Traceback (most recent call last):
  File "/home/c0rn3j/test.py", line 39, in <module>
    webnovel.signin(account, password)
  File "/home/c0rn3j/.local/lib/python3.9/site-packages/webnovel/bot.py", line 137, in signin
    captcha = self.driver.find_element_by_css_selector('#google-code-html iframe')
  File "/home/c0rn3j/.local/lib/python3.9/site-packages/selenium/webdriver/remote/webdriver.py", line 598, in find_element_by_css_selector
    return self.find_element(by=By.CSS_SELECTOR, value=css_selector)
  File "/home/c0rn3j/.local/lib/python3.9/site-packages/selenium/webdriver/remote/webdriver.py", line 976, in find_element
    return self.execute(Command.FIND_ELEMENT, {
  File "/home/c0rn3j/.local/lib/python3.9/site-packages/selenium/webdriver/remote/webdriver.py", line 321, in execute
    self.error_handler.check_response(response)
  File "/home/c0rn3j/.local/lib/python3.9/site-packages/selenium/webdriver/remote/errorhandler.py", line 242, in check_response
    raise exception_class(message, screen, stacktrace)
selenium.common.exceptions.NoSuchElementException: Message: no such element: Unable to locate element: {"method":"css selector","selector":"#google-code-html iframe"}
  (Session info: chrome=88.0.4324.182)
```

**Comment by m-haisham** [comment]
Ah. that is so weird it doesnt raise an error for me. I'll change it to an alternative method.

**Comment by m-haisham** [comment]
fixed 57e30f390abec1b8168b27543b8bff42983e582c

As per the documentation it is indeed supposed to throw an exception... no idea why it didnt for me.

Nice catch

**Comment by C0rn3j** [comment]
```python
Traceback (most recent call last):
  File "/home/c0rn3j/test.py", line 39, in <module>
    webnovel.signin(account, password)
  File "/home/c0rn3j/.local/lib/python3.9/site-packages/webnovel/bot.py", line 149, in signin
    raise GuardException
webnovel.exceptions.GuardException
```
There still seems to be an issue regarding redirection when the Guard challenge is triggered.

I get redirected to this mangled URL, which gives a 'not found' HTML page.
https://passport.webnovel.com/guard.htmlhttps://passport.webnovel.com/emaillogin.html&sessionKey=[key]&encry=[key]&email=[email]

**Comment by m-haisham** [comment]
That is the usual guard html.

Assuming manual is set to false,  this behaviors is expected. we can only check if guard is present when we are redirected to it.

**Comment by C0rn3j** [comment]
>That is the usual guard html.

But it looks broken? Unless the webnovel.com webserver does some really weird parsing.

```
https://passport.webnovel.com/guard.htmlhttps://passport.webnovel.com/emaillogin.html
                                       ^
```

>Assuming manual is set to false

My bad, I read the changes and somehow read that `manual` defaults to `True`. It was indeed not set(False).
I set it now.

This is the kind of popup I'd expect when manual is set to True:

![image](https://user-images.githubusercontent.com/1641362/109059036-0ac16680-76e4-11eb-97ba-00c3e3fc4b27.png)

But instead I get this form:
![image](https://user-images.githubusercontent.com/1641362/109058509-5c1d2600-76e3-11eb-8243-51faae56ca24.png)

Which even when filled out redirects back to the seemingly malformed URL:
![image](https://user-images.githubusercontent.com/1641362/109059427-8a4f3580-76e4-11eb-9808-734c59b39632.png)

**Comment by m-haisham** [comment]
Indeed. I'll take a look at it tomorrow.

**Comment by m-haisham** [comment]
fixed the guard redirect issue d63b8e0583a16f00b29fcde9d4c2d5229a681235

![Untitled](https://user-images.githubusercontent.com/47662901/109123600-d85b4c00-776b-11eb-935e-bf5accb239f5.png)

Do let me know if you find anything else

**Comment by C0rn3j** [comment]
![image](https://user-images.githubusercontent.com/1641362/109124992-d7b2bd80-774b-11eb-956f-732fe98c6e96.png)

Now I get stuck here, no captcha or guard were present for this login.

It does actually log in, but does not redirect back to a working page aftewards.

**Comment by m-haisham** [comment]
fixed 59cf7f184d042b8fb29490f7bdfa44ca19fd27d4

Also added [GuardHandler](https://github.com/mHaisham/webnovelbot/blob/master/webnovel/handlers/guard.py), this should help with automating interaction with Guard Page e979e323a35cbc2bb2195334ef1619df95608c59

Read more [here](https://github.com/mHaisham/webnovelbot/tree/master/webnovel/handlers)

**Comment by C0rn3j** [comment]
That works beautifully! I did not get hit by reCAPTCHA this time but the Guard integration is perfect.

Thanks a lot for fixing all of the problems!

I'll open a new Issue if I run into more problems as this one is cramped enough as is.
